### PR TITLE
upload template enhancement

### LINKED ISSUE #328: Upload OVF enhancement for known issues
This issue is for a fix track for the known issues for uploading OVF/OVA listed below. But this fix should be after pull request https://github.com/vmware/go-vcloud-director/pull/323 merged.
Here is a working fix commit in my forked repo
https://github.com/guo1017138/go-vcloud-director/commit/9d57a8f598123e9beb845b610d2cc8e8a60127fc 

Any comments is welcome.

Issue list:
1. In some bad situation, such like user offered OVF file doesn't contains ovf:size for files. The error as below.
Err: unhandled API response, please report this issue, status code: 416 Content-Range not satisfiable: Request invalid
2. If user's OVF has wrong ovf:size, in some condition, vCloud API returned 400 error without body, the error can't be populated correctly. It reports:
Err: API Error: 0:
But real response like below which body part is empty:
Status:400 Client sent more data than expected StatusCode:400 Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[Content-Length:[0]] Body:0xc00041c500 ContentLength:0 TransferEncoding:[] Close:false Uncompressed:false Trailer:map[] Request:0xc0003c8800 TLS:0xc00042c2c0
3. If I call ShowUploadProgress to show progress, it may go into infinity loop and never return.
Known triggers:
    * The upload task was cancelled by user on GUI or any error occur during show upload progress
    * ovf:size is not correct for vmdk. (rainy day case, may never match 100.00)

### PR DESCRIPTION
**Pull request by guo1017138**
Support upload ovf which has no ovf:size specified for each dependency file
ShowUploadProgress can exit whenever task stop, such like cancelled on GUI

Issue still exists: if ovf:size is bigger than real vmdk file size, upload task will hung there and never stop. I don't know how to fix

Close #328 

Signed-off-by: Guo, Larry (NSB - CN/Qingdao) <larry.guo@nokia-sbell.com>

### CODE REVIEW

**Comment by Didainius** [comment]
@guo1017138 , 
Sorry for ignoring you - we have been busy releasing [terraform-provider-vcd 3.0.0](https://registry.terraform.io/providers/vmware/vcd/latest).
Could you give an attempt to merge latest master and we will try to work through it to get it merged. Although it shows only changelog conflict - in fact it is quite hard to review as the diff is big.

Thanks!

**Comment by guo1017138** [comment]
@Didainius , Thank you. I merged with latest code and there is no conflict now. Please help to review.

File: govcd/catalog_test.go [hunk 1]
```diff
@@ -435,7 +444,7 @@ func (vcd *TestVCD) Test_CatalogUploadMediaImage_ShowUploadProgress_works(check
 
 	AddToCleanupList(itemName, "mediaCatalogImage", vcd.org.Org.Name+"|"+vcd.config.VCD.Catalog.Name, "Test_CatalogUploadMediaImage_ShowUploadProgress_works")
 
-	check.Assert(string(result), Matches, ".*Upload progress 100.00%")
+	check.Assert(string(result), Matches, ".*Upload progress 100.00%.*")
```
Comment (Didainius) [inline govcd/catalog_test.go:447]: I get errors in these two tests - 

```
START: catalog_test.go:419: TestVCD.Test_CatalogUploadMediaImage_ShowUploadProgress_works
Running: TestVCD.Test_CatalogUploadMediaImage_ShowUploadProgress_works
catalog_test.go:447:
    check.Assert(string(result), Matches, ".*Upload progress 100.00%.*")
... value string = "Waiting...\rUpload progress 0.00%\rUpload progress 100.00%\n"
... regex string = ".*Upload progress 100.00%.*"

FAIL: catalog_test.go:419: TestVCD.Test_CatalogUploadMediaImage_ShowUploadProgress_works
```


```
START: catalog_test.go:202: TestVCD.Test_UploadOvf_ShowUploadProgress_works
Running: TestVCD.Test_UploadOvf_ShowUploadProgress_works
catalog_test.go:230:
    check.Assert(string(result), Matches, ".*Upload progress 100.00%")
... value string = "Waiting...\rUpload progress 0.00%\rUpload progress 100.00%\n"
... regex string = ".*Upload progress 100.00%"

FAIL: catalog_test.go:202: TestVCD.Test_UploadOvf_ShowUploadProgress_works
```

Comment (guo1017138) [reply govcd/catalog_test.go]: Sorry about that. My vCloud ENV has some issue so I didn't run test completely. Per [the article here](https://github.com/google/re2/wiki/Syntax), dot in regex doesn't mathc new line \n by default. So, I enabled new line matching with s flag in the new code.

Comment (Didainius) [reply govcd/catalog_test.go]: Np - thanks. I will try to run some tests and see if it doesn't break any existing stuff.

**Comment by Didainius** [comment]
One more:
```
START: catalog_test.go:298: TestVCD.Test_UploadOvf_withoutVMDKSize
Running: TestVCD.Test_UploadOvf_withoutVMDKSize
catalog_test.go:302:
    checkUploadOvf(vcd, check, vcd.config.OVA.OvaWithoutSizePath, vcd.config.VCD.Catalog.Name, TestUploadOvf+"8")
catalog_test.go:323:
    check.Assert(err, IsNil)
... value *errors.errorString = &errors.errorString{s:"file size didn't match described in ovf: /var/folders/h2/c60z324s6219c3hc_w28hbx80000gp/T/govcd237219830/vm-20b41fa7-3212-4faf-9295-045d0bc49783-disk-0.vmdk. Unpacked files for checking are accessible in: /var/folders/h2/c60z324s6219c3hc_w28hbx80000gp/T/govcd237219830"} ("file size didn't match described in ovf: /var/folders/h2/c60z324s6219c3hc_w28hbx80000gp/T/govcd237219830/vm-20b41fa7-3212-4faf-9295-045d0bc49783-disk-0.vmdk. Unpacked files for checking are accessible in: /var/folders/h2/c60z324s6219c3hc_w28hbx80000gp/T/govcd237219830")

FAIL: catalog_test.go:298: TestVCD.Test_UploadOvf_withoutVMDKSize
```

OVA config:
```
ova:
  ovaPath: ../test-resources/test_vapp_template.ova
  ovaChunkedPath: ../test-resources/template_with_custom_chunk_size.ova
  ovaMultiVmPath: ../test-resources/vapp_with_3_vms.ova
  ovaWithoutSizePath: ../test-resources/template_without_vmdk_size.ova
``

**Comment by guo1017138** [comment]
> 
> 
> One more:
> 
> ```
> START: catalog_test.go:298: TestVCD.Test_UploadOvf_withoutVMDKSize
> Running: TestVCD.Test_UploadOvf_withoutVMDKSize
> catalog_test.go:302:
>     checkUploadOvf(vcd, check, vcd.config.OVA.OvaWithoutSizePath, vcd.config.VCD.Catalog.Name, TestUploadOvf+"8")
> catalog_test.go:323:
>     check.Assert(err, IsNil)
> ... value *errors.errorString = &errors.errorString{s:"file size didn't match described in ovf: /var/folders/h2/c60z324s6219c3hc_w28hbx80000gp/T/govcd237219830/vm-20b41fa7-3212-4faf-9295-045d0bc49783-disk-0.vmdk. Unpacked files for checking are accessible in: /var/folders/h2/c60z324s6219c3hc_w28hbx80000gp/T/govcd237219830"} ("file size didn't match described in ovf: /var/folders/h2/c60z324s6219c3hc_w28hbx80000gp/T/govcd237219830/vm-20b41fa7-3212-4faf-9295-045d0bc49783-disk-0.vmdk. Unpacked files for checking are accessible in: /var/folders/h2/c60z324s6219c3hc_w28hbx80000gp/T/govcd237219830")
> 
> FAIL: catalog_test.go:298: TestVCD.Test_UploadOvf_withoutVMDKSize
> ```
> 
> OVA config:
> 
> ```
> ova:
>   ovaPath: ../test-resources/test_vapp_template.ova
>   ovaChunkedPath: ../test-resources/template_with_custom_chunk_size.ova
>   ovaMultiVmPath: ../test-resources/vapp_with_3_vms.ova
>   ovaWithoutSizePath: ../test-resources/template_without_vmdk_size.ova
> ``
> ```

Thank you. It breaks in ova upload while my local ENV are using OVF file only.
But I am wondering why the travis-ci can pass each time? I supposed it shall run the test cased defined one by one.

**Comment by Didainius** [comment]
> > One more:
> > ```
> > START: catalog_test.go:298: TestVCD.Test_UploadOvf_withoutVMDKSize
> > Running: TestVCD.Test_UploadOvf_withoutVMDKSize
> > catalog_test.go:302:
> >     checkUploadOvf(vcd, check, vcd.config.OVA.OvaWithoutSizePath, vcd.config.VCD.Catalog.Name, TestUploadOvf+"8")
> > catalog_test.go:323:
> >     check.Assert(err, IsNil)
> > ... value *errors.errorString = &errors.errorString{s:"file size didn't match described in ovf: /var/folders/h2/c60z324s6219c3hc_w28hbx80000gp/T/govcd237219830/vm-20b41fa7-3212-4faf-9295-045d0bc49783-disk-0.vmdk. Unpacked files for checking are accessible in: /var/folders/h2/c60z324s6219c3hc_w28hbx80000gp/T/govcd237219830"} ("file size didn't match described in ovf: /var/folders/h2/c60z324s6219c3hc_w28hbx80000gp/T/govcd237219830/vm-20b41fa7-3212-4faf-9295-045d0bc49783-disk-0.vmdk. Unpacked files for checking are accessible in: /var/folders/h2/c60z324s6219c3hc_w28hbx80000gp/T/govcd237219830")
> > 
> > FAIL: catalog_test.go:298: TestVCD.Test_UploadOvf_withoutVMDKSize
> > ```
> > 
> > 
> > OVA config:
> > ```
> > ova:
> >   ovaPath: ../test-resources/test_vapp_template.ova
> >   ovaChunkedPath: ../test-resources/template_with_custom_chunk_size.ova
> >   ovaMultiVmPath: ../test-resources/vapp_with_3_vms.ova
> >   ovaWithoutSizePath: ../test-resources/template_without_vmdk_size.ova
> > ``
> > ```
> 
> Thank you. It breaks in ova upload while my local ENV are using OVF file only.
> But I am wondering why the travis-ci can pass each time? I supposed it shall run the test cased defined one by one.

Sadly travis does not run tests on real env (integration or whatever you call it) because there is on publicly available environment f VCD. It only does some static checks.

**Comment by guo1017138** [comment]
@Didainius Thanks. Got it.

**Comment by lvirbalas** [comment]
@guo1017138 are you planning to come back with an update, need to know what to do with this PR. Thanks!

**Comment by guo1017138** [comment]
> 
> 
> @guo1017138 are you planning to come back with an update, need to know what to do with this PR. Thanks!

Hi @lvirbalas , I didn't notice it conflict again. Now I resolved the conflict and please help to review.

**Review by Didainius — APPROVED** [review]
My testing passed. Thanks

**Review by dataclouder — CHANGES_REQUESTED** [review]
Some minor changes needed

**Review by lvirbalas — APPROVED** [review]
LGTM, thank you for your effort!

**Review by vbauzysvmware — APPROVED** [review]
LGTM. Thank you

**Review by dataclouder — APPROVED** [review]

**Comment by Didainius** [comment]
@guo1017138 ,
Thanks for PR. It is now merged and has a tag of `v2.10.0-alpha.3` before we release official one.

**Pull request merged**